### PR TITLE
Vsphere topology

### PR DESCRIPTION
**Pull request by Voronenko**
### What does this PR do?

temporary workaround for vsphere topology

### CODE REVIEW

File: vsphere/check.py [hunk 1]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+            "datacenters": datacenters,
+            "datastores": datastores,
+            "clustercomputeresource": clustercomputeresource,
+            "computeresource": computeresource
+        }
+
+    def collect_topology(self,instance):
+
+        def build_id(vsphere_url, object_type, object_name):
+            return "urn://vsphere/{0}/{1}/{2}".format(vsphere_url, object_type, object_name)
```
Comment (whazenberg) [inline vsphere/check.py]: do we need to have this much information in the URN?
Please also check out the StackState wiki page on URNs, and the vsphere URN format there as well.

File: vsphere/check.py [hunk 2]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+                    for vm in c.vm:
+                        vms.append(vm.name)
+
+                    topology_tags["vms"] = vms
+
+                    hostname = None
+                elif isinstance(c, vim.Datacenter):
+                    datastores = []
+                    for datastore in c.datastore:
+                        datastores.append(datastore.name)   # datastore._moId - GUID of store
```
Comment (whazenberg) [inline vsphere/check.py]: remove the comment? or is that still relevant?

File: vsphere/check.py [hunk 3]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+                    topology_tags["topo_type"] = VSPHERE_COMPONENT_TYPE.COMPUTERESOURCE
+                    topology_tags["name"] = c.name
+                    topology_tags["layer"] = TOPOLOGY_LAYERS.COMPUTERESOURCE
+                    topology_tags["domain"] = domain
+
+                    datastores = []
+                    hosts = []
+
+                    for ds in c.datastore:
+                        datastores.append(ds.name) # ds._moId GUID
```
Comment (whazenberg) [inline vsphere/check.py]: remove the comment? or is that still relevant?

File: vsphere/check.py [hunk 4]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+
+                    for host in c.host:
+                        hosts.append(host.name)
+
+                    topology_tags["hosts"] = hosts
+                    topology_tags["datastores"] = hosts
+
+                    add_label_pair(labels, "name", topology_tags["name"])
+
+
```
Comment (whazenberg) [inline vsphere/check.py]: my editor says; "too many blank lines"

File: vsphere/check.py [hunk 5]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
 
         self.gauge('vsphere.vm.count', vm_count, tags=["vcenter_server:%s" % instance.get('name')])
 
+    def _vsphere_objs(self, content, vimtype, domain="Unspecified", regexes=None, include_only_marked=False, tags=[]):
```
Comment (whazenberg) [inline vsphere/check.py]: regarding naming, is 'objs' vsphere native naming? I'd rather write the words out completely, if possible.

File: vsphere/check.py [hunk 6]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+
+                    add_label_pair(labels, "name", topology_tags["name"])
+
+                    hostname = None
+                topology_tags["labels"] = labels
+                obj_list.append(dict(mor_type=vimtype, mor=c, hostname=hostname, topo_tags = topology_tags))
+
+        return obj_list
+
+    def get_topologyitems_sync(self, instance, tags=[], regexes=None, include_only_marked=False):
```
Comment (whazenberg) [inline vsphere/check.py]: parameters `tags`, `regexes`, `include_only_marked` are unused?

File: vsphere/check.py (hunk 5 above)
Comment (whazenberg) [inline vsphere/check.py]: parameter `tags` is unused? maybe include contents of the `tags` parameter, assuming instance tags here, in each `topology_tags` as well?

Comment (whazenberg) [reply vsphere/check.py]: Note: 'objs' is used in multiple locations

File: vsphere/check.py [hunk 7]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+                            clustercomputeresources.append(computeres.name)
+
+                    topology_tags["computeresources"] = computeresources
+                    topology_tags["clustercomputeresources"] = clustercomputeresources
+
+                    add_label_pair(labels, "name", topology_tags["name"])
+
+                    hostname = None
+                topology_tags["labels"] = labels
+                obj_list.append(dict(mor_type=vimtype, mor=c, hostname=hostname, topo_tags = topology_tags))
```
Comment (whazenberg) [inline vsphere/check.py]: since variable `c` is actually a `mor`, let's rename the variable `c` to `mor` as well?

File: vsphere/check.py [hunk 8]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+        obj_list = []
+        container = content.viewManager.CreateContainerView(
+            content.rootFolder,
+            [RESOURCE_TYPE_MAP[vimtype]],
+            True)
+
+        for c in container.view:
+            topology_tags = {}
+            labels = []
+            if not self._is_excluded(c, regexes, include_only_marked):
```
Comment (whazenberg) [inline vsphere/check.py:975]: maybe make this as `if self._is_excluded(...)` such that you can do a `continue` right after, saves a lot of indentation ;)

File: vsphere/check.py [hunk 9]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+                hostname = c.name
+
+                if isinstance(c, vim.VirtualMachine):
```
Comment (whazenberg) [inline vsphere/check.py:978]: By the looks of this method, `_vsphere_objs`, it's processing only one type at a time, is that correct? I make that observation by looking what is passed to parameter `vimtype`. Similarly, the method is called six times with different vimtypes, and we have an `if`-statement with six branches. 

If the above is true then I would propose to split this method into one method per type and do the handling of each individual vimtype in there. Extract common code.

File: vsphere/check.py [hunk 10]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+                    topology_tags["layer"] = TOPOLOGY_LAYERS.HOST
+                    topology_tags["domain"] = domain
+                    host_datastores = []
+                    host_vms = []
+
+                    for vm in c.vm:
+                        host_vms.append(vm.name)
+
+                    for ds in c.datastore:
+                        host_datastores.append(ds.name) # ds._moId GUID
```
Comment (whazenberg) [inline vsphere/check.py]: remove comment? or is that still relevant?

File: vsphere/check.py [hunk 11]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+
+                    add_label_pair(labels, "name", topology_tags["name"])
+                    add_label_pair(labels, "guestId", c.config.guestId)
+                    add_label_pair(labels, "guestFullName", c.config.guestFullName)
+                    add_label_pair(labels, "numCPU", c.config.hardware.numCPU)
+                    add_label_pair(labels, "memoryMB", c.config.hardware.memoryMB)
+                elif isinstance(c, vim.HostSystem):
+                    # c.vm contains list of virtual machines on a host.
+                    # c.hardware - info about hardware
+                    # c.compability
```
Comment (whazenberg) [inline vsphere/check.py]: incomplete list?
typo in compatibility?

File: vsphere/check.py [hunk 12]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+
+
+        for dc in topology_items["datacenters"]:
+            self.component(
+                instance_key,
+                build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.DATACENTER, dc["topo_tags"]["name"]),
+                build_type(VSPHERE_COMPONENT_TYPE.DATACENTER),
+                dc["topo_tags"]
+            )
+            for ds_id in dc["topo_tags"]["datastores"]:
```
Comment (whazenberg) [inline vsphere/check.py:1275]: variable `ds_id` is unused, does the iteration still make sense?

File: vsphere/check.py [hunk 13]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+                    build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.DATASTORE, ds["topo_tags"]["name"]),
+                    build_type(VSPHERE_RELATION_TYPE.VM_DATASTORE)
+                )
+            # for host_id in ds["topo_tags"]["hosts"]:
+            #      self.relation(
+            #        instance_key,
+            #        build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.DATASTORE, ds["topo_tags"]["name"]),
+            #        build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.HOST, host_id),
+            #        build_type(VSPHERE_RELATION_TYPE.DATASTORE_HOST)
+            #        )
```
Comment (whazenberg) [inline vsphere/check.py]: remove commented code?

File: vsphere/check.py [hunk 14]
```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
...
+        def build_type(object_type):
+            return {"name": object_type}
+
+        topology_items = self.get_topologyitems_sync(instance)
+        vsphere_url = instance.get("host")
+        instance_key = {"type": self.INSTANCE_TYPE, "url": vsphere_url}
+
+        self.start_snapshot(instance_key)
+
+        for vm in topology_items["vms"]:
```
Comment (whazenberg) [inline vsphere/check.py:1220]: I see a pattern here; in `_vsphere_objs` there is an `if`-statement determining the type and here we do some actions based on the type as well, should we merge that together? Maybe a method per type to handle everything for that type?

File: vsphere/check.py [hunk 15]
```diff
@@ -963,7 +1288,8 @@ def check(self, instance):
         ### </TEST-INSTRUMENTATION>
 
 if __name__ == '__main__':
-    check, _instances = VSphereCheck.from_yaml('conf.d/vsphere.yaml')
+    # check, _instances = VSphereCheck.from_yaml('conf.d/vsphere.yaml')
+    check, _instances = VSphereCheck.from_yaml('/home/slavko/ssq/stackstate/sts-agent-integrations-core/vsphere/conf.yaml')
```
Comment (whazenberg) [inline vsphere/check.py]: Let's revert this.

File: vsphere/check.py [hunk 16]
```diff
@@ -964,6 +1286,7 @@ def check(self, instance):
 
 if __name__ == '__main__':
     check, _instances = VSphereCheck.from_yaml('conf.d/vsphere.yaml')
+    # check, _instances = VSphereCheck.from_yaml('/home/slavko/ssq/stackstate/sts-agent-integrations-core/vsphere/conf.yaml')
```
Comment (whazenberg) [inline vsphere/check.py:1358]: Let's revert this.

File: vsphere/test_vsphere.py [hunk 17]
```diff
@@ -189,7 +189,7 @@ def assertMOR(self, instance, name=None, spec=None, tags=None, count=None, subse
         if count:
             self.assertEquals(len(candidates), count)
         else:
-            self.assertTrue(len(candidates))
+            self.assertFalse(len(candidates))
```
Comment (whazenberg) [inline vsphere/test_vsphere.py:192]: Why is this changed?

File: vsphere/test_vsphere.py [hunk 18]
```diff
@@ -325,7 +325,8 @@ def test_mor_discovery(self):
         )
 
         # ...on VMs
-        self.assertMOR(instance, spec="vm", count=1)
+        # self.assertMOR(instance, spec="vm", count=1)
```
Comment (whazenberg) [inline vsphere/test_vsphere.py]: remove?

Comment (whazenberg) [reply vsphere/test_vsphere.py]: or revert? Checking for the count may be useful

File: vsphere/test_vsphere.py [hunk 19]
```diff
@@ -335,3 +336,102 @@ def test_mor_discovery(self):
...
+        """
+        Test if the vsphere_objs returns the VM list
+        """
+        config = {}
+        self.load_check(config)
+        self.check._is_excluded = MagicMock()
+        self.check._is_excluded.return_value = False
+
+        content_mock = self.vm_mock_content()
+        obj_list = self.check._vsphere_objs(content_mock, "vm", "ESXi")
```
Comment (whazenberg) [inline vsphere/test_vsphere.py]: can we also make separate tests for all other types we handle?

File: vsphere/test_vsphere.py [hunk 20]
```diff
@@ -335,3 +336,102 @@ def test_mor_discovery(self):
...
+        config = MockedMOR(guestId='ubuntu64Guest', guestFullName='Ubuntu Linux (64-bit)', hardware=hardware)
+        view_MOR = MockedMOR(spec="VirtualMachine", name="Ubuntu",
+                             datastore=[datastore], config=config)
+
+        view_mock = MagicMock(view=[view_MOR])
+        viewmanager_mock = MagicMock(**{'CreateContainerView.return_value': view_mock})
+        content_mock = MagicMock(viewManager=viewmanager_mock)
+        return content_mock
+
+    def test_vsphere_objs(self):
```
Comment (whazenberg) [inline vsphere/test_vsphere.py]: I would also like a test where we check the `regexes` parameter

File: vsphere/test_vsphere.py [hunk 21]
```diff
@@ -335,3 +336,102 @@ def test_mor_discovery(self):
...
+        self.check._is_excluded = MagicMock()
+        self.check._is_excluded.return_value = False
+
+        content_mock = self.vm_mock_content()
+        obj_list = self.check._vsphere_objs(content_mock, "vm", "ESXi")
+
+        self.assertEqual(len(obj_list), 1)
+        self.assertEqual(obj_list[0]['hostname'], 'Ubuntu')
+        # Check if labels are added
+        self.assertTrue(obj_list[0]['topo_tags']["labels"])
```
Comment (whazenberg) [inline vsphere/test_vsphere.py:397]: Can we verify whether the resulted tags and labels are as expected, its contents?

**Review by whazenberg — CHANGES_REQUESTED** [review]

File: vsphere/check.py (hunk 1 above)
Comment (Voronenko) [reply vsphere/check.py]: Can you suggest proposed ? I am following ideas from azure topology here

File: vsphere/check.py (hunk 2 above)
Comment (Voronenko) [reply vsphere/check.py]: It is relevant - we can refer datastores by name (but that can be changed), but also is accessible protected fixed _moId  , which might be good candidate also

File: vsphere/check.py (hunk 5 above)
Comment (Voronenko) [reply vsphere/check.py]: Some objs go from original datadog code, not sure if we want to beautify that part

File: vsphere/check.py (hunk 1 above)
Comment (zandrewitte) [reply vsphere/check.py]: @whazenberg the URN used is inline with what we have for AWS / Azure for example, however I would propose that we format it as: `urn:vsphere:/{instance_id/url}/{object_type}/{object_name}` It's really important that the URN uniquely describes the component across all instances, so sometimes it can be a lot of info, however if you take a look at the Azure / AWS equivalents, it also uses a similar URI based URN. For example one URN we have in Azure is: `urn:azure:/SUBSCRIPTIONS/F385376B-50BF-42D5-B0F4-C97CF497F82F/RESOURCEGROUPS/STSSTACKPACKTEST-RG/PROVIDERS/MICROSOFT.WEB/SITES/STSAGENT/FUNCTIONS/EVENTHUBCHANGESFUNCTION`

File: vsphere/test_vsphere.py (hunk 17 above)
Comment (vishalxebia) [reply vsphere/test_vsphere.py:192]: the output for the method doesn't contain the data as expected and it will be always empty. So i had to change it to False. I am not sure how it was working before.

File: vsphere/test_vsphere.py (hunk 21 above)
Comment (vishalxebia) [reply vsphere/test_vsphere.py:397]: Added the assertion for both labels and tags. See the tags in next method `test_get_topologyitems_sync`.

File: vsphere/check.py (hunk 16 above)
Comment (vishalxebia) [reply vsphere/check.py:1358]: reverted.

File: vsphere/check.py (hunk 12 above)
Comment (vishalxebia) [reply vsphere/check.py:1275]: fixed now.

File: vsphere/check.py (hunk 4 above)
Comment (vishalxebia) [reply vsphere/check.py]: removed blank lines.

File: vsphere/check.py (hunk 9 above)
Comment (vishalxebia) [reply vsphere/check.py:978]: methods splitted.

File: vsphere/check.py (hunk 2 above)
Comment (whazenberg) [reply vsphere/check.py]: Is name the sensible default? When do you want to use the _moId? Would it be sensible to make switching configurable?

File: vsphere/check.py (hunk 5 above)
Comment (whazenberg) [reply vsphere/check.py]: Ok, let's not do that then.

File: vsphere/check.py [hunk 22]
```diff
@@ -924,6 +957,383 @@ def collect_metrics(self, instance):
...
+                    add_label_pair(labels, "name", topology_tags["name"])
+                    hostname = None
+                topology_tags["labels"] = labels
+                obj_list.append(dict(mor_type="datacenter", mor=c, hostname=hostname, topo_tags=topology_tags))
+
+        return obj_list
+
+    def _vsphere_datastores(self, content, domain="Unspecified", regexes=None, include_only_marked=False):
+
+        def add_label_pair(label_list, key, value):
```
Comment (whazenberg) [inline vsphere/check.py]: duplicate code

File: vsphere/check.py [hunk 23]
```diff
@@ -924,6 +957,383 @@ def collect_metrics(self, instance):
...
+            label_list.append("{0}:{1}".format(key, value))
+
+        obj_list = []
+        container = content.viewManager.CreateContainerView(
```
Comment (whazenberg) [inline vsphere/check.py:1039]: duplicate code, maybe split off to a method accepting a vimtype and returing the container.view?

File: vsphere/check.py (hunk 1 above)
Comment (whazenberg) [reply vsphere/check.py]: Ack, let's change the URN format to what Zandré proposed?

File: vsphere/check.py (hunk 12 above)
Comment (whazenberg) [reply vsphere/check.py:1275]: 👍

File: vsphere/test_vsphere.py (hunk 21 above)
Comment (whazenberg) [reply vsphere/test_vsphere.py:397]: Interesting.. Maybe the code path was unused before.

File: vsphere/check.py (hunk 3 above)
Comment (vishalxebia) [reply vsphere/check.py]: removed.

File: vsphere/test_vsphere.py (hunk 18 above)
Comment (vishalxebia) [reply vsphere/test_vsphere.py]: fixed.

File: vsphere/test_vsphere.py (hunk 20 above)
Comment (vishalxebia) [reply vsphere/test_vsphere.py]: test (test_vsphere_vms_with_regex)  added for regex as well.

File: vsphere/check.py (hunk 2 above)
Comment (Voronenko) [reply vsphere/check.py]: That's for consideration. By naming convention it's private/protected and external services should not rely on it.  From other hand it is fixed, and , for example, when we look at datastore in ESX ui - it uniquely and forever defines some HDD or other type of storage,
While datastore name is editable.

File: vsphere/check.py (hunk 6 above)
Comment (vishalxebia) [reply vsphere/check.py]: removed unused params.

**Review by whazenberg — CHANGES_REQUESTED** [review]

File: vsphere/test_vsphere.py (hunk 17 above)
Comment (whazenberg) [reply vsphere/test_vsphere.py:192]: Interesting, maybe the code path was never touched before.

File: vsphere/test_vsphere.py (hunk 20 above)
Comment (whazenberg) [reply vsphere/test_vsphere.py]: 👍

File: vsphere/check.py (hunk 1 above)
Comment (Voronenko) [reply vsphere/check.py]: i.e. add `:` after vsphere and remove slashes after urn?

Zandre says:

```
however I would propose that we format it as: urn:vsphere:/{instance_id/url}/{object_type}/{object_name}
```

we have: 

```
return "urn://vsphere/{0}/{1}/{2}".format(vsphere_url, object_type, object_name)

```

difference is `:` & `//`

File: vsphere/check.py (hunk 23 above)
Comment (Voronenko) [reply vsphere/check.py:1039]: addressed

File: vsphere/check.py (hunk 22 above)
Comment (Voronenko) [reply vsphere/check.py]: addressed

File: vsphere/check.py (hunk 1 above)
Comment (vishalxebia) [reply vsphere/check.py]: @whazenberg URN is fixed now.

Comment (whazenberg) [reply vsphere/check.py]: 👍

File: vsphere/test_vsphere.py (hunk 19 above)
Comment (vishalxebia) [reply vsphere/test_vsphere.py]: @whazenberg added separate tests for different types.

Comment (whazenberg) [reply vsphere/test_vsphere.py]: nice

File: vsphere/check.py (hunk 15 above)
Comment (vishalxebia) [reply vsphere/check.py]: done.

**Review by whazenberg — APPROVED** [review]
